# Data Ingestion & Wrangling -- Montgomery County Vision Zero Traffic Progress
Data 205 - Spring 2021  
Mary Glantz

#### Import the libraries needed to import and wrangle the data

In [31]:
import json
import numpy as np
import pandas as pd
import requests



#### Use API's to import data from three Montgomery county datasets (driver info, collision info, and non-motorist collision info) into dataframes.


In [137]:
driverdf = pd.DataFrame(requests.get('https://data.montgomerycountymd.gov/resource/mmzv-x632.json').json())
driverdf.head()

report_number local_case_number                agency_name  \
0    DD5647000F         210007208  Rockville Police Departme   
1    DD5647000F         210007208  Rockville Police Departme   
2   MCP102200FH         210007202   Montgomery County Police   
3   MCP102200FH         210007202   Montgomery County Police   
4    DD5647000C         210007203  Rockville Police Departme   

        acrs_report_type          crash_date_time        route_type  \
0  Property Damage Crash  2021-02-25T10:15:00.000  Maryland (State)   
1  Property Damage Crash  2021-02-25T10:15:00.000  Maryland (State)   
2           Injury Crash  2021-02-25T09:16:00.000  Maryland (State)   
3           Injury Crash  2021-02-25T09:16:00.000  Maryland (State)   
4  Property Damage Crash  2021-02-25T09:16:00.000  Maryland (State)   

          road_name cross_street_type cross_street_name municipality  ...  \
0        NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
1        NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
2          RIDGE RD            County          BRINK RD          N/A  ...   
3          RIDGE RD            County          BRINK RD          N/A  ...   
4  W MONTGOMERY AVE      Municipality         NELSON ST    ROCKVILLE  ...   

  :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9     latitude  \
0                           3                          25  39.09043867   
1                           3                          25  39.09043867   
2                           2                          39  39.21003667   
3                           2                          39  39.21003667   
4                           3                          25  39.08605617   

      longitude :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
0  -77.12867933                          71                          12   
1  -77.12867933                          71                          12   
2      -77.2399                          36                           1   
3      -77.2399                          36                           1   
4  -77.17062833                          71                          12   

                                         geolocation off_road_description  \
0  {'latitude': '39.09043867', 'longitude': '-77....                  NaN   
1  {'latitude': '39.09043867', 'longitude': '-77....                  NaN   
2  {'latitude': '39.21003667', 'longitude': '-77....                  NaN   
3  {'latitude': '39.21003667', 'longitude': '-77....                  NaN   
4  {'latitude': '39.08605617', 'longitude': '-77....                  NaN   

  related_non_motorist non_motorist_substance_abuse  
0                  NaN                          NaN  
1                  NaN                          NaN  
2                  NaN                          NaN  
3                  NaN                          NaN  
4                  NaN                          NaN  

[5 rows x 50 columns]

In [4]:
incidentsdf = pd.DataFrame(requests.get('https://data.montgomerycountymd.gov/resource/bhju-22kf.json').json())
incidentsdf.head()

report_number local_case_number                agency_name  \
0    DD5647000F         210007208  Rockville Police Departme   
1   MCP102200FH         210007202   Montgomery County Police   
2    DD5647000C         210007203  Rockville Police Departme   
3   MCP3036001R         210007162   Montgomery County Police   
4   MCP32180018         210007164   Montgomery County Police   

        acrs_report_type          crash_date_time hit_run        route_type  \
0  Property Damage Crash  2021-02-25T10:15:00.000      No  Maryland (State)   
1           Injury Crash  2021-02-25T09:16:00.000      No  Maryland (State)   
2  Property Damage Crash  2021-02-25T09:16:00.000      No  Maryland (State)   
3  Property Damage Crash  2021-02-24T20:56:00.000      No  Maryland (State)   
4  Property Damage Crash  2021-02-24T20:25:00.000      No            County   

  mile_point mile_point_direction lane_direction  ...  \
0      24.09                 East           West  ...   
1       0.54                North          South  ...   
2      21.29                 East           West  ...   
3       3.76                North          South  ...   
4       2.28                 East           East  ...   

  :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9  \
0                           3                          25   
1                           2                          39   
2                           3                          25   
3                           3                          36   
4                           4                          23   

  :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
0                          71                          12   
1                          36                           1   
2                          71                          12   
3                          35                          21   
4                          17                           1   

  :@computed_region_a9cs_3ed7 :@computed_region_r648_kzwt        lane_type  \
0                           3                           3              NaN   
1                           2                           5              NaN   
2                           3                           3              NaN   
3                           3                           5  RIGHT TURN LANE   
4                           4                           4              NaN   

  off_road_description related_non_motorist non_motorist_substance_abuse  
0                  NaN                  NaN                          NaN  
1                  NaN                  NaN                          NaN  
2                  NaN                  NaN                          NaN  
3                  NaN                  NaN                          NaN  
4                  NaN                  NaN                          NaN  

[5 rows x 51 columns]

In [5]:
nonmotoristdf = pd.DataFrame(requests.get('https://data.montgomerycountymd.gov/resource/n7fk-dce5.json').json())
nonmotoristdf.head()

report_number local_case_number                agency_name acrs_report_type  \
0    DD56370028         210006695  Rockville Police Departme     Injury Crash   
1   MCP3209002B         210005989   Montgomery County Police     Injury Crash   
2   MCP3274000F         210005911   Montgomery County Police     Injury Crash   
3   MCP229800VL         210005324   Montgomery County Police     Injury Crash   
4   MCP3289000G         210005250   Montgomery County Police     Injury Crash   

           crash_date_time                   off_road_description  \
0  2021-02-21T16:23:00.000  PARKING GARAGE OF 1628 E JEFFERSON ST   
1  2021-02-15T20:20:00.000                                    NaN   
2  2021-02-15T10:49:00.000                                    NaN   
3  2021-02-10T15:16:00.000                                    NaN   
4  2021-02-10T06:35:00.000                                    NaN   

  related_non_motorist           collision_type  weather           light  ...  \
0           PEDESTRIAN                    OTHER    CLEAR             N/A  ...   
1           PEDESTRIAN                    OTHER  RAINING            DUSK  ...   
2           PEDESTRIAN                    OTHER      N/A        DAYLIGHT  ...   
3            BICYCLIST  STRAIGHT MOVEMENT ANGLE   CLOUDY        DAYLIGHT  ...   
4           PEDESTRIAN           SINGLE VEHICLE   CLOUDY  DARK LIGHTS ON  ...   

  :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
0                          60                          12   
1                        3065                           1   
2                         111                           1   
3                          39                           1   
4                          21                           1   

  :@computed_region_a9cs_3ed7 :@computed_region_r648_kzwt        route_type  \
0                           3                           3               NaN   
1                           5                           1            County   
2                           1                           3            County   
3                           2                           5            County   
4                           5                           1  Maryland (State)   

            road_name cross_street_type cross_street_name municipality  \
0                 NaN               NaN               NaN          NaN   
1           WAYNE AVE        US (State)       GEORGIA AVE          N/A   
2            BROAD ST            County      GREENLAWN DR          N/A   
3  FATHER HURLEY BLVD            County       WISTERIA DR          N/A   
4   UNIVERSITY BLVD E            County   WILLIAMSBURG DR          N/A   

  surface_condition  
0               NaN  
1               WET  
2               DRY  
3               DRY  
4               DRY  

[5 rows x 39 columns]


#### Begin cleaning driver info dataframe  
First, examine the information about the dataframe.


In [138]:
driverdf.shape

(1000, 50)

In [139]:
driverdf.columns

Index(['report_number', 'local_case_number', 'agency_name', 'acrs_report_type',
       'crash_date_time', 'route_type', 'road_name', 'cross_street_type',
       'cross_street_name', 'municipality', 'collision_type', 'weather',
       'surface_condition', 'light', 'traffic_control',
       'driver_substance_abuse', 'person_id', 'driver_at_fault',
       'injury_severity', 'circumstance', 'driver_distracted_by',
       'drivers_license_state', 'vehicle_id', 'vehicle_damage_extent',
       'vehicle_first_impact_location', 'vehicle_second_impact_location',
       'vehicle_body_type', 'vehicle_movement', 'vehicle_continuing_dir',
       'vehicle_going_dir', 'speed_limit', 'driverless_vehicle',
       'parked_vehicle', ':@computed_region_a9cs_3ed7', 'vehicle_year',
       'vehicle_make', ':@computed_region_r648_kzwt', 'vehicle_model',
       ':@computed_region_vu5j_pcmz', 'equipment_problems',
       ':@computed_region_tx5f_5em3', ':@computed_region_kbsp_ykn9',
       'latitude', 'longitude'

In [140]:
driverdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 50 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   report_number                   1000 non-null   object
 1   local_case_number               1000 non-null   object
 2   agency_name                     1000 non-null   object
 3   acrs_report_type                1000 non-null   object
 4   crash_date_time                 1000 non-null   object
 5   route_type                      902 non-null    object
 6   road_name                       906 non-null    object
 7   cross_street_type               902 non-null    object
 8   cross_street_name               906 non-null    object
 9   municipality                    906 non-null    object
 10  collision_type                  1000 non-null   object
 11  weather                         1000 non-null   object
 12  surface_condition               906 non-null    o

The number of nun-null entries indicates whether or not all 1000 entries are non-null.  In several columns, there are fewer than 1000 non-null entries, which means there should be nulls.  The next step is to check for consistent missing value notation.

#### Convert date-time objects to datetime format

In [141]:
driverdf['crash_date_time'] = pd.to_datetime(driverdf['crash_date_time'])
driverdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 50 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   report_number                   1000 non-null   object        
 1   local_case_number               1000 non-null   object        
 2   agency_name                     1000 non-null   object        
 3   acrs_report_type                1000 non-null   object        
 4   crash_date_time                 1000 non-null   datetime64[ns]
 5   route_type                      902 non-null    object        
 6   road_name                       906 non-null    object        
 7   cross_street_type               902 non-null    object        
 8   cross_street_name               906 non-null    object        
 9   municipality                    906 non-null    object        
 10  collision_type                  1000 non-null   object        
 11  weath

#### Check for consistency of variables within columns
I'm not going to check road names because of the expected massive variability within those categories.

In [142]:
reports = driverdf.report_number
reports_unique = reports.unique()
reports.nunique()

595

This is very interesting, because it indicates that of the 1000 entries, there are only 595 unique reports.  I will check that by using "duplicated" to see how many duplicate entries there are.

In [143]:
duplicateRowsDF = driverdf[driverdf.duplicated('report_number')]
duplicateRowsDF

report_number local_case_number                agency_name  \
1      DD5647000F         210007208  Rockville Police Departme   
3     MCP102200FH         210007202   Montgomery County Police   
6     MCP3036001R         210007162   Montgomery County Police   
8     MCP32180018         210007164   Montgomery County Police   
10     DD5595009H         210007147  Rockville Police Departme   
..            ...               ...                        ...   
988   MCP30390055         210001767   Montgomery County Police   
990   MCP2927003R         210002381   Montgomery County Police   
992   MCP1128005B         210002369   Montgomery County Police   
995   MCP31740031         210002359   Montgomery County Police   
998   MCP2671002J         210002364   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
3             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
6    Property Damage Crash 2021-02-24 20:56:00  Maryland (State)   
8    Property Damage Crash 2021-02-24 20:25:00            County   
10            Injury Crash 2021-02-24 18:25:00  Maryland (State)   
..                     ...                 ...               ...   
988  Property Damage Crash 2021-01-19 16:00:00  Maryland (State)   
990  Property Damage Crash 2021-01-19 15:52:00  Maryland (State)   
992  Property Damage Crash 2021-01-19 15:36:00  Maryland (State)   
995           Injury Crash 2021-01-19 15:28:00            County   
998  Property Damage Crash 2021-01-19 15:18:00            County   

                  road_name cross_street_type    cross_street_name  \
1                NORBECK RD      Municipality            E GUDE DR   
3                  RIDGE RD            County             BRINK RD   
6    MONTGOMERY VILLAGE AVE           Unknown  SPUR TO RUSSELL AVE   
8               RANDOLPH RD  Maryland (State)      CONNECTICUT AVE   
10         W MONTGOMERY AVE      Municipality            NELSON ST   
..                      ...               ...                  ...   
988           EAST WEST HWY  Maryland (State)        COLESVILLE RD   
990       UNIVERSITY BLVD E            County     INDIAN SPRING DR   
992          ROCKVILLE PIKE            County          SECURITY LA   
995         WATKINS MILL RD            County       APPLE RIDGE RD   
998               FENTON ST  Maryland (State)     PHILADELPHIA AVE   

    municipality  ... :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9  \
1      ROCKVILLE  ...                           3                          25   
3            N/A  ...                           2                          39   
6            N/A  ...                           3                          36   
8            N/A  ...                           4                          23   
10     ROCKVILLE  ...                           3                          25   
..           ...  ...                         ...                         ...   
988          N/A  ...                           5                           7   
990          N/A  ...                           5                           8   
992          N/A  ...                           1                          19   
995          N/A  ...                           2                          38   
998          N/A  ...                           5                           7   

        latitude     longitude :@computed_region_d7bw_bq6x  \
1    39.09043867  -77.12867933                          71   
3    39.21003667      -77.2399                          36   
6    39.15403167  -77.20922667                          35   
8     39.0572375  -77.07452383                          17   
10   39.08578501  -77.17066705                          71   
..           ...           ...                         ...   
988  38.99352352  -77.03212171                        3065   
990  39.01405612  -77.00419546                          21   
992  39.04210833  -77.

So approximately half of my entries are duplicates.  I will need to figure out what I want to do about that.  I will probably look at the duplicates and see what the differences between the entries are, and then delete or merge the rows as appropriate.



In [144]:
cases = driverdf.local_case_number
cases_unique = cases.unique()
reports.nunique()

595

In [145]:
duplicateCasesDF = driverdf[driverdf.duplicated('local_case_number')]
duplicateCasesDF

report_number local_case_number                agency_name  \
1      DD5647000F         210007208  Rockville Police Departme   
3     MCP102200FH         210007202   Montgomery County Police   
6     MCP3036001R         210007162   Montgomery County Police   
8     MCP32180018         210007164   Montgomery County Police   
10     DD5595009H         210007147  Rockville Police Departme   
..            ...               ...                        ...   
988   MCP30390055         210001767   Montgomery County Police   
990   MCP2927003R         210002381   Montgomery County Police   
992   MCP1128005B         210002369   Montgomery County Police   
995   MCP31740031         210002359   Montgomery County Police   
998   MCP2671002J         210002364   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
3             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
6    Property Damage Crash 2021-02-24 20:56:00  Maryland (State)   
8    Property Damage Crash 2021-02-24 20:25:00            County   
10            Injury Crash 2021-02-24 18:25:00  Maryland (State)   
..                     ...                 ...               ...   
988  Property Damage Crash 2021-01-19 16:00:00  Maryland (State)   
990  Property Damage Crash 2021-01-19 15:52:00  Maryland (State)   
992  Property Damage Crash 2021-01-19 15:36:00  Maryland (State)   
995           Injury Crash 2021-01-19 15:28:00            County   
998  Property Damage Crash 2021-01-19 15:18:00            County   

                  road_name cross_street_type    cross_street_name  \
1                NORBECK RD      Municipality            E GUDE DR   
3                  RIDGE RD            County             BRINK RD   
6    MONTGOMERY VILLAGE AVE           Unknown  SPUR TO RUSSELL AVE   
8               RANDOLPH RD  Maryland (State)      CONNECTICUT AVE   
10         W MONTGOMERY AVE      Municipality            NELSON ST   
..                      ...               ...                  ...   
988           EAST WEST HWY  Maryland (State)        COLESVILLE RD   
990       UNIVERSITY BLVD E            County     INDIAN SPRING DR   
992          ROCKVILLE PIKE            County          SECURITY LA   
995         WATKINS MILL RD            County       APPLE RIDGE RD   
998               FENTON ST  Maryland (State)     PHILADELPHIA AVE   

    municipality  ... :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9  \
1      ROCKVILLE  ...                           3                          25   
3            N/A  ...                           2                          39   
6            N/A  ...                           3                          36   
8            N/A  ...                           4                          23   
10     ROCKVILLE  ...                           3                          25   
..           ...  ...                         ...                         ...   
988          N/A  ...                           5                           7   
990          N/A  ...                           5                           8   
992          N/A  ...                           1                          19   
995          N/A  ...                           2                          38   
998          N/A  ...                           5                           7   

        latitude     longitude :@computed_region_d7bw_bq6x  \
1    39.09043867  -77.12867933                          71   
3    39.21003667      -77.2399                          36   
6    39.15403167  -77.20922667                          35   
8     39.0572375  -77.07452383                          17   
10   39.08578501  -77.17066705                          71   
..           ...           ...                         ...   
988  38.99352352  -77.03212171                        3065   
990  39.01405612  -77.00419546                          21   
992  39.04210833  -77.

I have the same situation with cases, so I assume they are the same---I'll double check by checking both columns.

In [146]:
driverdf[driverdf['local_case_number']=='210002364']

report_number local_case_number               agency_name  \
997   MCP2671002J         210002364  Montgomery County Police   
998   MCP2671002J         210002364  Montgomery County Police   

          acrs_report_type     crash_date_time route_type  road_name  \
997  Property Damage Crash 2021-01-19 15:18:00     County  FENTON ST   
998  Property Damage Crash 2021-01-19 15:18:00     County  FENTON ST   

    cross_street_type cross_street_name municipality  ...  \
997  Maryland (State)  PHILADELPHIA AVE          N/A  ...   
998  Maryland (State)  PHILADELPHIA AVE          N/A  ...   

    :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9     latitude  \
997                           5                           7  38.98793019   
998                           5                           7  38.98793019   

       longitude :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
997  -77.0235645                        3065                           1   
998  -77.0235645                        3065                           1   

                                           geolocation off_road_description  \
997  {'latitude': '38.98793019', 'longitude': '-77....                  NaN   
998  {'latitude': '38.98793019', 'longitude': '-77....                  NaN   

    related_non_motorist non_motorist_substance_abuse  
997                  NaN                          NaN  
998                  NaN                          NaN  

[2 rows x 50 columns]

In [147]:
driverdf[driverdf['local_case_number']=='210002359']

report_number local_case_number               agency_name  \
994   MCP31740031         210002359  Montgomery County Police   
995   MCP31740031         210002359  Montgomery County Police   

    acrs_report_type     crash_date_time route_type        road_name  \
994     Injury Crash 2021-01-19 15:28:00     County  WATKINS MILL RD   
995     Injury Crash 2021-01-19 15:28:00     County  WATKINS MILL RD   

    cross_street_type cross_street_name municipality  ...  \
994            County    APPLE RIDGE RD          N/A  ...   
995            County    APPLE RIDGE RD          N/A  ...   

    :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9     latitude  \
994                           2                          38  39.18419833   
995                           2                          38  39.18419833   

        longitude :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
994  -77.21089333                          28                           1   
995  -77.21089333                          28                           1   

                                           geolocation off_road_description  \
994  {'latitude': '39.18419833', 'longitude': '-77....                  NaN   
995  {'latitude': '39.18419833', 'longitude': '-77....                  NaN   

    related_non_motorist non_motorist_substance_abuse  
994                  NaN                          NaN  
995                  NaN                          NaN  

[2 rows x 50 columns]

In these test cases, it appears all of the information matches, and the records are duplicates.  I will remove all duplicates.

In [148]:
driverdf.drop_duplicates(subset=['report_number', 'local_case_number'], keep='last')   # I'm keeping the last rather than the first just in case the last has more recent data.

report_number local_case_number                agency_name  \
1      DD5647000F         210007208  Rockville Police Departme   
3     MCP102200FH         210007202   Montgomery County Police   
4      DD5647000C         210007203  Rockville Police Departme   
6     MCP3036001R         210007162   Montgomery County Police   
8     MCP32180018         210007164   Montgomery County Police   
..            ...               ...                        ...   
993   MCP271700TB         210002365   Montgomery County Police   
995   MCP31740031         210002359   Montgomery County Police   
996   MCP3095002K         210002355   Montgomery County Police   
998   MCP2671002J         210002364   Montgomery County Police   
999   MCP3074004L         210002362   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
3             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
4    Property Damage Crash 2021-02-25 09:16:00  Maryland (State)   
6    Property Damage Crash 2021-02-24 20:56:00  Maryland (State)   
8    Property Damage Crash 2021-02-24 20:25:00            County   
..                     ...                 ...               ...   
993           Injury Crash 2021-01-19 15:31:00  Maryland (State)   
995           Injury Crash 2021-01-19 15:28:00            County   
996  Property Damage Crash 2021-01-19 15:20:00            County   
998  Property Damage Crash 2021-01-19 15:18:00            County   
999  Property Damage Crash 2021-01-19 15:18:00               NaN   

                  road_name cross_street_type    cross_street_name  \
1                NORBECK RD      Municipality            E GUDE DR   
3                  RIDGE RD            County             BRINK RD   
4          W MONTGOMERY AVE      Municipality            NELSON ST   
6    MONTGOMERY VILLAGE AVE           Unknown  SPUR TO RUSSELL AVE   
8               RANDOLPH RD  Maryland (State)      CONNECTICUT AVE   
..                      ...               ...                  ...   
993         CONNECTICUT AVE      Municipality       ROSEMARY ST #2   
995         WATKINS MILL RD            County       APPLE RIDGE RD   
996               ELDRID DR            County          JOHNSON AVE   
998               FENTON ST  Maryland (State)     PHILADELPHIA AVE   
999                     NaN               NaN                  NaN   

       municipality  ... :@computed_region_tx5f_5em3  \
1         ROCKVILLE  ...                           3   
3               N/A  ...                           2   
4         ROCKVILLE  ...                           3   
6               N/A  ...                           3   
8               N/A  ...                           4   
..              ...  ...                         ...   
993  CHEVY CHASE #4  ...                           1   
995             N/A  ...                           2   
996             N/A  ...                           5   
998             N/A  ...                           5   
999             NaN  ...                           5   

    :@computed_region_kbsp_ykn9     latitude     longitude  \
1                            25  39.09043867  -77.12867933   
3                            39  39.21003667      -77.2399   
4                            25  39.08605617  -77.17062833   
6                            36  39.15403167  -77.20922667   
8                            23   39.0572375  -77.07452383   
..                          ...          ...           ...   
993                           5  38.97779433  -77.07710317   
995                          38  39.18419833  -77.21089333   
996                          26     39.06622  -76.99723667   
998                           7  38.98793019   -77.0235645   
999                           8    39.000365  -77.00984667   

    :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
1                            71                          12   
3        

Now I will clean up the remaining data.  First checking for entries that need to be merged/changed.  Anytime they use different terms for the same thing (N/A, unknown, etc.), I will replace with one consistent term.

In [149]:
agency_name = driverdf.agency_name
agency_uniques = agency_name.unique()
agency_uniques

array(['Rockville Police Departme', 'Montgomery County Police',
       'Maryland-National Capital', 'Gaithersburg Police Depar',
       'Takoma Park Police Depart'], dtype=object)

This has all consistent terms, but I need to correct the typos caused by shortening the names.

In [150]:
# Using replace by dictionary to correct the names

driverdf['agency_name'] = driverdf['agency_name'].replace({'Rockville Police Departme':'Rockville Police', 'Maryland-National Capital':'Maryland-National Capital Police', 
                                                     'Gaithersburg Police Depar':'Gaithersburg Police', 'Takoma Park Police Depart': 'Takoma Park Police'})
agency_uniques = agency_name.unique()
agency_uniques

array(['Rockville Police', 'Montgomery County Police',
       'Maryland-National Capital Police', 'Gaithersburg Police',
       'Takoma Park Police'], dtype=object)

In [151]:
acrs_report_type = driverdf.acrs_report_type
acrs_uniques = acrs_report_type.unique()
acrs_uniques

array(['Property Damage Crash', 'Injury Crash'], dtype=object)

In [152]:
route_type = driverdf.route_type
route_uniques = route_type.unique()
route_uniques

array(['Maryland (State)', 'County', nan, 'Municipality', 'US (State)',
       'Interstate (State)', 'Other Public Roadway', 'Ramp', 'Government'],
      dtype=object)

There is only one type of unknown here (nan), so I will leave it.

In [153]:
cross_street_type = driverdf.cross_street_type
cross_street_uniques = cross_street_type.unique()
cross_street_uniques

array(['Municipality', 'County', 'Unknown', 'Maryland (State)', nan,
       'US (State)', 'Ramp', 'Other Public Roadway', 'Service Road',
       'Interstate (State)'], dtype=object)

'Unknown' and 'nan' should be combined.

In [154]:
driverdf['cross_street_type'] = driverdf['cross_street_type'].fillna('Unknown')
cross_street_type = driverdf.cross_street_type
cross_street_uniques = cross_street_type.unique()
cross_street_uniques

array(['Municipality', 'County', 'Unknown', 'Maryland (State)',
       'US (State)', 'Ramp', 'Other Public Roadway', 'Service Road',
       'Interstate (State)'], dtype=object)

In [155]:
municipality = driverdf.municipality
municipality_uniques = municipality.unique()
municipality_uniques

array(['ROCKVILLE', 'N/A', nan, 'GAITHERSBURG', 'SOMERSET', 'KENSINGTON',
       'TAKOMA PARK', 'POOLESVILLE', 'FRIENDSHIP HEIGHTS',
       'CHEVY CHASE #4'], dtype=object)

'N/A' and 'nan' should be combined.

In [156]:
driverdf['municipality'] = driverdf['municipality'].fillna('N/A')
municipality = driverdf.municipality
municipality_uniques = municipality.unique()
municipality_uniques

array(['ROCKVILLE', 'N/A', 'GAITHERSBURG', 'SOMERSET', 'KENSINGTON',
       'TAKOMA PARK', 'POOLESVILLE', 'FRIENDSHIP HEIGHTS',
       'CHEVY CHASE #4'], dtype=object)

In [157]:
collision_type = driverdf.collision_type
collision_type_uniques = collision_type.unique()
collision_type_uniques

array(['SAME DIRECTION SIDESWIPE', 'STRAIGHT MOVEMENT ANGLE',
       'SINGLE VEHICLE', 'SAME DIR REAR END', 'HEAD ON LEFT TURN',
       'OTHER', 'HEAD ON', 'SAME DIRECTION RIGHT TURN',
       'OPPOSITE DIRECTION SIDESWIPE', 'UNKNOWN',
       'ANGLE MEETS LEFT HEAD ON', 'ANGLE MEETS RIGHT TURN',
       'OPPOSITE DIR BOTH LEFT TURN', 'ANGLE MEETS LEFT TURN',
       'SAME DIR BOTH LEFT TURN', 'SAME DIRECTION LEFT TURN',
       'SAME DIR REND RIGHT TURN', 'SAME DIR REND LEFT TURN', 'N/A'],
      dtype=object)

In [161]:
# Replace 'N/A' with 'UNKNOWN'
driverdf['collision_type'] = driverdf['collision_type'].replace({'N/A': 'UNKNOWN'})
collision_type = driverdf.collision_type
collision_type_uniques = collision_type.unique()
collision_type_uniques


array(['SAME DIRECTION SIDESWIPE', 'STRAIGHT MOVEMENT ANGLE',
       'SINGLE VEHICLE', 'SAME DIR REAR END', 'HEAD ON LEFT TURN',
       'OTHER', 'HEAD ON', 'SAME DIRECTION RIGHT TURN',
       'OPPOSITE DIRECTION SIDESWIPE', 'UNKNOWN',
       'ANGLE MEETS LEFT HEAD ON', 'ANGLE MEETS RIGHT TURN',
       'OPPOSITE DIR BOTH LEFT TURN', 'ANGLE MEETS LEFT TURN',
       'SAME DIR BOTH LEFT TURN', 'SAME DIRECTION LEFT TURN',
       'SAME DIR REND RIGHT TURN', 'SAME DIR REND LEFT TURN'],
      dtype=object)

In [166]:
weather = driverdf.weather
weather_uniques = weather.unique()
weather_uniques

array(['CLEAR', 'UNKNOWN', 'CLOUDY', 'RAINING', 'SLEET', 'SNOW',
       'WINTRY MIX', 'OTHER', 'BLOWING SNOW', 'FOGGY'], dtype=object)

In [167]:
# Replace 'N/A' with 'UNKNOWN'  

driverdf['weather'] = driverdf['weather'].replace({'N/A':'UNKNOWN'})
weather = driverdf.weather
weather_uniques = weather.unique()
weather_uniques


array(['CLEAR', 'UNKNOWN', 'CLOUDY', 'RAINING', 'SLEET', 'SNOW',
       'WINTRY MIX', 'OTHER', 'BLOWING SNOW', 'FOGGY'], dtype=object)

In [168]:
surface_condition = driverdf.surface_condition
surface_condition_uniques = surface_condition.unique()
surface_condition_uniques

array(['DRY', 'WET', 'UNKNOWN', 'SNOW', 'SLUSH', 'ICE',
       'WATER(STANDING/MOVING)'], dtype=object)

I will combine "nan", "N/A", and "UNKNOWN", and replace nan with "UNKNOWN".

In [169]:
driverdf['surface_condition'] = driverdf['surface_condition'].replace({'N/A': 'UNKNOWN'})
driverdf['surface_condition'] = driverdf['surface_condition'].fillna('UNKNOWN')
surface_condition = driverdf.surface_condition
surface_condition_uniques = surface_condition.unique()
surface_condition_uniques

array(['DRY', 'WET', 'UNKNOWN', 'SNOW', 'SLUSH', 'ICE',
       'WATER(STANDING/MOVING)'], dtype=object)

In [172]:
light = driverdf.light
light_uniques = light.unique()
light_uniques

array(['DAYLIGHT', 'DARK LIGHTS ON', 'DUSK', 'DARK NO LIGHTS', 'UNKNOWN',
       'DAWN', 'DARK -- UNKNOWN LIGHTING', 'OTHER'], dtype=object)

In [173]:
# Replace 'N/A' with 'UNKNOWN'
driverdf['light'] = driverdf['light'].replace({'N/A':'UNKNOWN'})
light = driverdf.light
light_uniques = light.unique()
light_uniques

array(['DAYLIGHT', 'DARK LIGHTS ON', 'DUSK', 'DARK NO LIGHTS', 'UNKNOWN',
       'DAWN', 'DARK -- UNKNOWN LIGHTING', 'OTHER'], dtype=object)

In [176]:
traffic_control = driverdf.traffic_control
traf_ctrl_uniques = traffic_control.unique()
traf_ctrl_uniques

array(['UNKNOWN', 'TRAFFIC SIGNAL', 'NO CONTROLS', 'STOP SIGN',
       'YIELD SIGN', 'FLASHING TRAFFIC SIGNAL', 'OTHER'], dtype=object)

In [178]:
# Replace 'N/A' with 'UNKNOWN'
driverdf['traffic_control'] = driverdf['traffic_control'].replace({'N/A': 'UNKNOWN'})
traffic_control = driverdf.traffic_control
traf_ctrl_uniques = traffic_control.unique()
traf_ctrl_uniques

array(['UNKNOWN', 'TRAFFIC SIGNAL', 'NO CONTROLS', 'STOP SIGN',
       'YIELD SIGN', 'FLASHING TRAFFIC SIGNAL', 'OTHER'], dtype=object)

In [85]:
substance_abuse = driverdf.driver_substance_abuse
substance_uniques = substance_abuse.unique()
substance_uniques


array(['N/A', 'NONE DETECTED', 'ALCOHOL PRESENT', 'UNKNOWN',
       'ALCOHOL CONTRIBUTED', 'OTHER', 'ILLEGAL DRUG PRESENT',
       'MEDICATION CONTRIBUTED', 'MEDICATION PRESENT',
       'COMBINATION CONTRIBUTED'], dtype=object)

In [179]:
# Replace "N/A" with "UNKNOWN"
driverdf['driver_substance_abuse'] = driverdf['driver_substance_abuse'].replace({'N/A': 'UNKNOWN'})
substance_abuse = driverdf.driver_substance_abuse
substance_uniques = substance_abuse.unique()
substance_uniques

array(['UNKNOWN', 'NONE DETECTED', 'ALCOHOL PRESENT',
       'ALCOHOL CONTRIBUTED', 'OTHER', 'ILLEGAL DRUG PRESENT',
       'MEDICATION CONTRIBUTED', 'MEDICATION PRESENT',
       'COMBINATION CONTRIBUTED'], dtype=object)

In [88]:
driver_at_fault = driverdf.driver_at_fault
at_fault_uniques = driver_at_fault.unique()
at_fault_uniques

array(['Yes', 'No', 'Unknown'], dtype=object)

In [89]:
injury_severity = driverdf.injury_severity 
injury_severity_uniques = injury_severity.unique()
injury_severity_uniques

array(['NO APPARENT INJURY', 'SUSPECTED MINOR INJURY', 'POSSIBLE INJURY',
       'SUSPECTED SERIOUS INJURY'], dtype=object)

In [90]:
circumstance = driverdf.circumstance
circumstance_uniques = circumstance.unique()
circumstance_uniques

array(['N/A',
       'PHYSICAL OBSTRUCTION(S), VISION OBSTRUCTION (INCL. BLINDED BY SUN)',
       'ANIMAL, N/A', 'ICY OR SNOW-COVERED, N/A, WET', 'N/A, WET',
       'ICY OR SNOW-COVERED, RAIN, SNOW', 'RAIN, SNOW, WET',
       'N/A, SLEET, HAIL, FREEZ. RAIN',
       'ICY OR SNOW-COVERED, SLEET, HAIL, FREEZ. RAIN',
       'SLEET, HAIL, FREEZ. RAIN, WET',
       'ICY OR SNOW-COVERED, RAIN, SNOW, SLEET, HAIL, FREEZ. RAIN',
       'N/A, VISION OBSTRUCTION (INCL. BLINDED BY SUN)',
       'ICY OR SNOW-COVERED, N/A',
       'V EXHAUST SYSTEM|R OTHER ROAD, V WIPERS|W OTHER ENVIRONMENTAL',
       'ICY OR SNOW-COVERED, SLEET, HAIL, FREEZ. RAIN, WET',
       'ICY OR SNOW-COVERED, RAIN, SNOW, SLEET, HAIL, FREEZ. RAIN, WET',
       'DEBRIS OR OBSTRUCTION, N/A',
       'ICY OR SNOW-COVERED, V WIPERS|W OTHER ENVIRONMENTAL',
       'ICY OR SNOW-COVERED, RAIN, SNOW, WET',
       'N/A, V EXHAUST SYSTEM|R OTHER ROAD', 'N/A, RAIN, SNOW',
       'RAIN, SNOW, V EXHAUST SYSTEM|R OTHER ROAD'], dtype=object)

In [91]:
driver_distracted_by = driverdf.driver_distracted_by
driver_distracted_by_uniques = driver_distracted_by.unique()
driver_distracted_by_uniques

array(['LOOKED BUT DID NOT SEE', 'NOT DISTRACTED',
       'INATTENTIVE OR LOST IN THOUGHT', 'UNKNOWN', 'OTHER DISTRACTION',
       'DISTRACTED BY OUTSIDE PERSON OBJECT OR EVENT',
       'BY MOVING OBJECT IN VEHICLE', 'NO DRIVER PRESENT',
       'EATING OR DRINKING', 'USING DEVICE OBJECT BROUGHT INTO VEHICLE',
       'OTHER CELLULAR PHONE RELATED',
       'OTHER ELECTRONIC DEVICE (NAVIGATIONAL PALM PILOT)',
       'TEXTING FROM A CELLULAR PHONE', 'DIALING CELLULAR PHONE',
       'BY OTHER OCCUPANTS'], dtype=object)

In [92]:
drivers_license_state = driverdf.drivers_license_state
drivers_license_state_uniques = drivers_license_state.unique()
drivers_license_state_uniques

array(['MD', 'CA', nan, 'DC', 'DE', 'NC', 'VA', 'XX', 'PA', 'CT', 'NJ',
       'NY', 'GA', 'WV', 'CO', 'AZ', 'MH', 'MA', 'FL', 'MI', 'SC', 'OH',
       'ME', 'RI', 'TN', 'LA', 'IL'], dtype=object)

In [180]:
# 'XX' appears to be the same as nan---I'll replace nan with 'XX'
driverdf['drivers_license_state'] = driverdf['drivers_license_state'].fillna('XX')
drivers_license_state = driverdf.drivers_license_state
drivers_license_state_uniques = drivers_license_state.unique()
drivers_license_state_uniques

array(['MD', 'CA', 'XX', 'DC', 'DE', 'NC', 'VA', 'PA', 'CT', 'NJ', 'NY',
       'GA', 'WV', 'CO', 'AZ', 'MH', 'MA', 'FL', 'MI', 'SC', 'OH', 'ME',
       'RI', 'TN', 'LA', 'IL'], dtype=object)

In [94]:
vehicle_damage_extent = driverdf.vehicle_damage_extent
vehicle_damage_extent_uniques = vehicle_damage_extent.unique()
vehicle_damage_extent_uniques

array(['DISABLING', 'FUNCTIONAL', 'SUPERFICIAL', 'DESTROYED', 'UNKNOWN',
       'NO DAMAGE', 'OTHER', 'N/A'], dtype=object)

In [181]:
driverdf['vehicle_damage_extent'] = driverdf['vehicle_damage_extent'].replace({'N/A':'UNKNOWN'})
vehicle_damage_extent = driverdf.vehicle_damage_extent
vehicle_damage_extent_uniques = vehicle_damage_extent.unique()
vehicle_damage_extent_uniques

array(['DISABLING', 'FUNCTIONAL', 'SUPERFICIAL', 'DESTROYED', 'UNKNOWN',
       'NO DAMAGE', 'OTHER'], dtype=object)

In [95]:
vehicle_first_impact_location = driverdf.vehicle_first_impact_location
vehicle_first_impact_location_uniques = vehicle_first_impact_location.unique()
vehicle_first_impact_location_uniques

array(['ONE OCLOCK', 'SEVEN OCLOCK', 'ELEVEN OCLOCK', 'TWELVE OCLOCK',
       'TWO OCLOCK', 'SIX OCLOCK', 'FIVE OCLOCK', 'EIGHT OCLOCK',
       'TEN OCLOCK', 'FOUR OCLOCK', 'THREE OCLOCK', 'ROOF TOP',
       'UNDERSIDE', 'NINE OCLOCK', 'UNKNOWN', 'NON-COLLISION'],
      dtype=object)

In [96]:
vehicle_second_impact_location = driverdf.vehicle_second_impact_location
vehicle_second_impact_location_uniques = vehicle_second_impact_location.unique()
vehicle_second_impact_location_uniques

array(['ONE OCLOCK', 'SEVEN OCLOCK', 'ELEVEN OCLOCK', 'TWELVE OCLOCK',
       'TWO OCLOCK', 'SIX OCLOCK', 'FIVE OCLOCK', 'EIGHT OCLOCK',
       'TEN OCLOCK', 'FOUR OCLOCK', 'THREE OCLOCK', 'ROOF TOP',
       'UNDERSIDE', 'NINE OCLOCK', 'UNKNOWN', 'NON-COLLISION'],
      dtype=object)

In [97]:
vehicle_body_type = driverdf.vehicle_body_type
vehicle_body_type_uniques = vehicle_body_type.unique()
vehicle_body_type_uniques

array(['PASSENGER CAR', 'PICKUP TRUCK', 'VAN', 'TRANSIT BUS',
       '(SPORT) UTILITY VEHICLE', 'FIRE VEHICLE/NON EMERGENCY',
       'SCHOOL BUS',
       'CARGO VAN/LIGHT TRUCK 2 AXLES (OVER 10,000LBS (4,536 KG))',
       'POLICE VEHICLE/EMERGENCY',
       'MEDIUM/HEAVY TRUCKS 3 AXLES (OVER 10,000LBS (4,536KG))',
       'POLICE VEHICLE/NON EMERGENCY', 'STATION WAGON',
       'OTHER LIGHT TRUCKS (10,000LBS (4,536KG) OR LESS)',
       'AMBULANCE/EMERGENCY', 'TRUCK TRACTOR', 'UNKNOWN',
       'FIRE VEHICLE/EMERGENCY', 'OTHER', nan, 'MOTORCYCLE', 'OTHER BUS',
       'RECREATIONAL VEHICLE', 'MOPED', 'AMBULANCE/NON EMERGENCY'],
      dtype=object)

In [182]:
# Replace nan with 'UNKNOWN'
driverdf['vehicle_body_type'] = driverdf['vehicle_body_type'].fillna('UNKNOWN')
vehicle_body_type = driverdf.vehicle_body_type
vehicle_body_type_uniques = vehicle_body_type.unique()
vehicle_body_type_uniques

array(['PASSENGER CAR', 'PICKUP TRUCK', 'VAN', 'TRANSIT BUS',
       '(SPORT) UTILITY VEHICLE', 'FIRE VEHICLE/NON EMERGENCY',
       'SCHOOL BUS',
       'CARGO VAN/LIGHT TRUCK 2 AXLES (OVER 10,000LBS (4,536 KG))',
       'POLICE VEHICLE/EMERGENCY',
       'MEDIUM/HEAVY TRUCKS 3 AXLES (OVER 10,000LBS (4,536KG))',
       'POLICE VEHICLE/NON EMERGENCY', 'STATION WAGON',
       'OTHER LIGHT TRUCKS (10,000LBS (4,536KG) OR LESS)',
       'AMBULANCE/EMERGENCY', 'TRUCK TRACTOR', 'UNKNOWN',
       'FIRE VEHICLE/EMERGENCY', 'OTHER', 'MOTORCYCLE', 'OTHER BUS',
       'RECREATIONAL VEHICLE', 'MOPED', 'AMBULANCE/NON EMERGENCY'],
      dtype=object)

In [98]:
vehicle_movement = driverdf.vehicle_movement
vehicle_movement_uniques = vehicle_movement.unique()
vehicle_movement_uniques

array(['MOVING CONSTANT SPEED', 'MAKING LEFT TURN', 'MAKING RIGHT TURN',
       'STARTING FROM PARKED', 'SLOWING OR STOPPING',
       'STOPPED IN TRAFFIC LANE', 'CHANGING LANES', 'ACCELERATING',
       'PARKING', 'BACKING', 'PARKED', 'STARTING FROM LANE', 'UNKNOWN',
       'SKIDDING', 'PASSING', 'OTHER', 'N/A', 'NEGOTIATING A CURVE',
       'MAKING U TURN', 'LEAVING TRAFFIC LANE', 'RIGHT TURN ON RED',
       'ENTERING TRAFFIC LANE'], dtype=object)

In [99]:
vehicle_continuing_dir = driverdf.vehicle_continuing_dir
vehicle_continuing_dir_uniques = vehicle_continuing_dir.unique()
vehicle_continuing_dir_uniques

array(['West', 'North', 'South', 'East', nan, 'Unknown'], dtype=object)

In [185]:
# Replace nan with 'UNKNOWN'
driverdf['vehicle_continuing_dir'] = driverdf['vehicle_continuing_dir'].fillna('UNKNOWN')
driverdf['vehicle_continuing_dir'] = driverdf['vehicle_continuing_dir'].replace({'Unknown':'UNKNOWN'})
vehicle_continuing_dir = driverdf.vehicle_continuing_dir
vehicle_continuing_dir_uniques = vehicle_continuing_dir.unique()
vehicle_continuing_dir_uniques

array(['West', 'North', 'South', 'East', 'UNKNOWN'], dtype=object)

In [100]:
vehicle_going_dir = driverdf.vehicle_going_dir
vehicle_going_dir_uniques = vehicle_going_dir.unique()
vehicle_going_dir_uniques

array(['West', 'East', 'South', 'North', nan, 'Unknown'], dtype=object)

In [186]:
# Replace nan with 'UNKNOWN'
driverdf['vehicle_going_dir'] = driverdf['vehicle_going_dir'].fillna('UNKNOWN')
driverdf['vehicle_going_dir'] = driverdf['vehicle_going_dir'].replace({'Unknown':'UNKNOWN'})
vehicle_going_dir = driverdf.vehicle_going_dir
vehicle_going_dir_uniques = vehicle_going_dir.unique()
vehicle_going_dir_uniques

array(['West', 'East', 'South', 'North', 'UNKNOWN'], dtype=object)

In [101]:
speed_limit = driverdf.speed_limit
speed_limit_uniques = speed_limit.unique()
speed_limit_uniques

array(['50', '40', '30', '35', '15', '25', '5', '45', '10', '0', '20',
       '55', '60'], dtype=object)

In [102]:
parked_vehicle = driverdf.parked_vehicle
parked_vehicle_uniques = parked_vehicle.unique()
parked_vehicle_uniques

array(['No', 'Yes'], dtype=object)

In [103]:
vehicle_year = driverdf.vehicle_year
vehicle_year_uniques = vehicle_year.unique()
vehicle_year_uniques

array(['2006', '2014', '2016', '2015', '2007', '2017', '2009', '2013',
       '2018', '2010', '1999', '1996', '2005', '2020', '2019', '2008',
       '2021', '2012', '2011', '2003', '1997', '2001', '1998', '2002',
       '2004', '2000', '1989', '0', '1986', '1901', '1992', '1995',
       '1990', '2022'], dtype=object)

In [104]:
vehicle_make = driverdf.vehicle_make
vehicle_make_uniques = vehicle_make.unique()
vehicle_make_uniques

array(['TOYOTA', 'TOYT', 'FORD', 'BMW', 'NISSAN', 'HONDA', 'NEW FLYER',
       'MERZ', 'MAZDA', 'INFI', 'HYUN', 'CHEV', 'SUBARU', 'HYUND',
       'THOMAS', 'CHEVROLET', 'JEEP', 'TRANSIT', 'LEXUS', 'CHEVY', 'MACK',
       'LEXU', 'HOND', 'VOLK', 'NISS', 'DODGE', 'FREIGHTLINER', 'KIA',
       'RANGE', 'HYUNDAI', 'DODG', 'NISAN', 'GMC', 'HYUD', 'MERCEDES',
       'ACURA', 'MERC', 'VOLVO', 'LINCOLN', 'INTERNATIONAL', 'AUDI',
       'MNNI', 'UNKNOWN', 'PRTA', 'KENWTH', 'KAWASAKI', 'MAZD', 'PTRB',
       'FRHT', 'RAM', 'NOVABUS', 'TBU', 'GILL', 'KENWORTH', 'MITSUBISHI',
       'MITSUBUSHI', 'BUICK', 'MASERATTI', 'HUYN', 'PIERCE', 'NEWFLYER',
       'VOLKWAGEN', 'INFINITI', 'SUBA', 'LANDOVER', 'TOYO', 'PETE',
       'ISUZU', 'MITS', 'CADI', 'PORSCHE', 'NWFL', 'INTL', 'CHEVORLET',
       'VW', 'SPARTAN', 'CHRY', 'MERCEDES BENZ', 'VOLV', 'TESLA',
       'SATURN', 'CHRYSLER', 'LANDROVER', 'LEXIS', 'VOLKWAGON',
       'VOLKSWAGEN', 'PIER', 'LAND ROVER', 'FRAIGHT LINER', 'NEW', 'SPAR',
       'HON

In [187]:
# Standardize make names --- unfortunately reviewing the make and model together reveals merc can be either mercedes or mercury...
# Some of these abbreviations are mysteries.
driverdf['vehicle_make'] = driverdf['vehicle_make'].replace({'TOYT': 'TOYOTA', 'MERZ': 'MERCEDES', 'INFI': 'INFINITI', 'HYUN': 'HYUNDAI', 'CHEV': 'CHEVROLET',
                                                             'HYUND': 'HYUNDAI', 'CHEVY': 'CHEVROLET', 'LEXU': 'LEXUS', 'HOND': 'HONDA', 'VOLK': 'VOLKSWAGEN',
                                                             'NISS': 'NISSAN', 'RANGE': 'RANGE_ROVER', 'DODG': 'DODGE', 'NISAN': 'NISSAN', 'HYUD': 'HYUNDAI',
                                                             'MNNI': 'MINI', 'MAZD': 'MAZDA', 'MITSUBUSHI': 'MITSUBISHI', 'HUYN': 'HYUNDAI', 'VOLKWAGEN': 'VOLKSWAGEN', 
                                                             'SUBA': 'SUBARU', 'LANDOVER': 'LANDROVER', 'TOYO': 'TOYOTA', 'MITS': 'MITSUBISHI', 'CADI': 'CADILLAC',
                                                             'CHEVORLET': 'CHEVROLET', 'VW': 'VOLKSWAGEN', 'CHRY':'CHRYSLER', 'VOLKWAGON': 'VOLKSWAGEN', 'LAND ROVER': 'LANDROVER', 
                                                             'FRAIGHT LINER': 'FREIGHTLINER', 'HONDM': 'HONDA', 'TOYTA': 'TOYOTA', 'SSUBARU': 'SUBARU', 'VOLKSWAGON': 'VOLKSWAGEN', 'XX': 'UNKNOWN',
                                                             'ACUR': 'ACURA', 'BUICCK': 'BUICK', 'HYNDAI': 'HYUNDAI', 'UNK': 'UNKNOWN', 'HYUNDA':'HYUNDAI', 'SUB': 'SUBARU',
                                                             'Unknown':'UNKNOWN'})
vehicle_make = driverdf.vehicle_make
vehicle_make_uniques = vehicle_make.unique()
vehicle_make_uniques

array(['TOYOTA', 'FORD', 'BMW', 'NISSAN', 'HONDA', 'NEW FLYER',
       'MERCEDES', 'MAZDA', 'INFINITI', 'CHEVROLET', 'HYUNDAI', 'SUBARU',
       'THOMAS', 'JEEP', 'TRANSIT', 'LEXUS', 'MACK', 'VOLKSWAGEN',
       'DODGE', 'FREIGHTLINER', 'KIA', 'RANGE_ROVER', 'GMC', 'ACURA',
       'MERC', 'VOLVO', 'INTERNATIONAL', 'LINCOLN', 'AUDI', 'MINI',
       'UNKNOWN', 'PRTA', 'KENWTH', 'KAWASAKI', 'PTRB', 'FRHT', 'RAM',
       'NOVABUS', 'TBU', 'GILL', 'KENWORTH', 'MITSUBISHI', 'BUICK',
       'MASERATTI', 'PIERCE', 'NEWFLYER', 'LANDROVER', 'PETE', 'ISUZU',
       'CADILLAC', 'PORSCHE', 'NWFL', 'INTL', 'SPARTAN', 'CHRYSLER',
       'MERCEDES BENZ', 'VOLV', 'TESLA', 'SATURN', 'LEXIS', 'PIER', 'NEW',
       'SPAR', 'GILG', 'LLOYDS', 'HINO', 'POSTAL TRUCK', 'STLG', 'VESPA',
       'PONTIAC', 'DIESEL URBAN TRANSIT'], dtype=object)

In [105]:
vehicle_model = driverdf.vehicle_model
vehicle_model_uniques = vehicle_model.unique()
vehicle_model_uniques

array(['4S', 'F-150', 'SIENNA', 'M4S', '328I', 'ALTIMA', 'CIVIC', 'TBU',
       'TK', 'COLORADO', 'ACCORD CROSSTOUR', 'CROSSTREK', 'ELANTRA',
       'CAMRY', 'SCHOOL BUS', 'VN', 'CRV', 'TUNDRA', 'TACOMA', 'SENTRA',
       'COROLLA', 'BUS', 'SEQUOIA', 'IS 300', 'SILVERADO', 'F150',
       'CAMARO', 'CN', '500', 'HR-V', 'ACCORD', 'JETTA', 'SUV', 'ODYSSEY',
       'JOURNEY', 'FOCUS', 'COBALT', 'EXPRESS', 'FORESTER', 'MALIBU',
       'CX9', 'TRUCK', 'SPECTRA', 'EVOQUE', 'OPTIMA', 'FORTE', '4D',
       'RAV4', 'JX35', 'CHARGER', 'ALTIMA 4S', 'PATRIOT TK', 'PATHFINDER',
       'CANYON', 'TUCSON', 'EXPLORER', 'LEGACY', 'C300L', 'TL', '4 DOOR',
       'C5500', 'ELEMENT', 'SW', 'ROGUE', 'TUNDA', '4RUNNER', 'MURANO',
       'VOLT', 'ESCAPE', 'WRANGLER', 'GLE350', 'SD', 'HRV', 'PILOT',
       'DUMP TRUCK', 'XC30', 'ML 350', 'EQUINOX LT', 'PRIUS', 'SQ5',
       'NAVIGATOR', 'SUBURBAN', 'COOPER', '1500', 'MUSTANG', 'M2 106',
       'PICK UP', 'UNKNOWN', 'BA', 'TITAN', 'WT 1500', 'RANGER',
       '4

In [106]:
equipment_problems = driverdf.equipment_problems
equipment_problems_uniques = equipment_problems.unique()
equipment_problems_uniques

array(['N/A', 'NO MISUSE', 'FACING WRONG WAY', 'UNKNOWN', 'OTHER'],
      dtype=object)

NOTE:  This column doesn't seem to include any information of interest, so I'll delete it.

In [109]:
related_non_motorist = driverdf.related_non_motorist
related_non_motorist_uniques = related_non_motorist.unique()
related_non_motorist_uniques

array([nan, 'PEDESTRIAN', 'BICYCLIST', 'MACHINE OPERATOR/RIDER'],
      dtype=object)

In [188]:
# Fill na's with "UNKNOWN"
driverdf['related_non_motorist'] = driverdf['related_non_motorist'].fillna("UNKNOWN")
related_non_motorist = driverdf.related_non_motorist
related_non_motorist_uniques = related_non_motorist.unique()
related_non_motorist_uniques

array(['UNKNOWN', 'PEDESTRIAN', 'BICYCLIST', 'MACHINE OPERATOR/RIDER'],
      dtype=object)

In [110]:
non_motorist_substance_abuse = driverdf.non_motorist_substance_abuse
non_motorist_substance_abuse_uniques = non_motorist_substance_abuse.unique()
non_motorist_substance_abuse_uniques

array([nan, 'NONE DETECTED', 'ALCOHOL PRESENT'], dtype=object)

In [189]:
# Fill na's with "UNKNOWN"
driverdf['non_motorist_substance_abuse'] = driverdf['non_motorist_substance_abuse'].fillna("UNKNOWN")
non_motorist_substance_abuse = driverdf.non_motorist_substance_abuse
non_motorist_substance_abuse_uniques = non_motorist_substance_abuse.unique()
non_motorist_substance_abuse_uniques

array(['UNKNOWN', 'NONE DETECTED', 'ALCOHOL PRESENT'], dtype=object)

Now that I see how the data in each column is labeled, I can change the labels for consistency.  (And delete the columns that don't have information I am interested in.

In [121]:
# Delete Columns I don't want to work with---I want to drop 13, but can only drop 8 per command so I will save in a temp

tempdriverdf = driverdf.drop(['person_id', 'vehicle_id', 'driverless_vehicle', ':@computed_region_a9cs_3ed7', ':@computed_region_r648_kzwt', ':@computed_region_vu5j_pcmz', 
              'equipment_problems'], axis=1)

report_number local_case_number                agency_name  \
0      DD5647000F         210007208  Rockville Police Departme   
1      DD5647000F         210007208  Rockville Police Departme   
2     MCP102200FH         210007202   Montgomery County Police   
3     MCP102200FH         210007202   Montgomery County Police   
4      DD5647000C         210007203  Rockville Police Departme   
..            ...               ...                        ...   
995   MCP31740031         210002359   Montgomery County Police   
996   MCP3095002K         210002355   Montgomery County Police   
997   MCP2671002J         210002364   Montgomery County Police   
998   MCP3074004L         210002362   Montgomery County Police   
999   MCP2671002J         210002364   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
0    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
2             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
3             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
4    Property Damage Crash 2021-02-25 09:16:00  Maryland (State)   
..                     ...                 ...               ...   
995           Injury Crash 2021-01-19 15:28:00            County   
996  Property Damage Crash 2021-01-19 15:20:00            County   
997  Property Damage Crash 2021-01-19 15:18:00            County   
998  Property Damage Crash 2021-01-19 15:18:00               NaN   
999  Property Damage Crash 2021-01-19 15:18:00            County   

            road_name cross_street_type cross_street_name municipality  ...  \
0          NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
1          NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
2            RIDGE RD            County          BRINK RD          N/A  ...   
3            RIDGE RD            County          BRINK RD          N/A  ...   
4    W MONTGOMERY AVE      Municipality         NELSON ST    ROCKVILLE  ...   
..                ...               ...               ...          ...  ...   
995   WATKINS MILL RD            County    APPLE RIDGE RD          N/A  ...   
996         ELDRID DR            County       JOHNSON AVE          N/A  ...   
997         FENTON ST  Maryland (State)  PHILADELPHIA AVE          N/A  ...   
998               NaN               NaN               NaN          NaN  ...   
999         FENTON ST  Maryland (State)  PHILADELPHIA AVE          N/A  ...   

    :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9     latitude  \
0                             3                          25  39.09043867   
1                             3                          25  39.09043867   
2                             2                          39  39.21003667   
3                             2                          39  39.21003667   
4                             3                          25  39.08605617   
..                          ...                         ...          ...   
995                           2                          38  39.18419833   
996                           5                          26     39.06622   
997                           5                           7  38.98793019   
998                           5                           8    39.000365   
999                           5                           7  38.98793019   

        longitude :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
0    -77.12867933                          71                          12   
1    -77.12867933                          71                          12   
2        -77.2399                          36                           1   
3        -77.2399                          36                           1   
4    -77.17062833                          71                          12   
..            ...                         ...                         

In [123]:
# I will now drop the 5 columns left in the tempdriver dataframe and save it in a new df called driver2df

driver2df = tempdriverdf.drop([':@computed_region_tx5f_5em3',':@computed_region_tx5f_5em3', ':@computed_region_kbsp_ykn9', ':@computed_region_d7bw_bq6x', ':@computed_region_rbt8_3x7n',
                               'off_road_description'], axis = 1)
driver2df



report_number local_case_number                agency_name  \
0      DD5647000F         210007208  Rockville Police Departme   
1      DD5647000F         210007208  Rockville Police Departme   
2     MCP102200FH         210007202   Montgomery County Police   
3     MCP102200FH         210007202   Montgomery County Police   
4      DD5647000C         210007203  Rockville Police Departme   
..            ...               ...                        ...   
995   MCP31740031         210002359   Montgomery County Police   
996   MCP3095002K         210002355   Montgomery County Police   
997   MCP2671002J         210002364   Montgomery County Police   
998   MCP3074004L         210002362   Montgomery County Police   
999   MCP2671002J         210002364   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
0    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
2             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
3             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
4    Property Damage Crash 2021-02-25 09:16:00  Maryland (State)   
..                     ...                 ...               ...   
995           Injury Crash 2021-01-19 15:28:00            County   
996  Property Damage Crash 2021-01-19 15:20:00            County   
997  Property Damage Crash 2021-01-19 15:18:00            County   
998  Property Damage Crash 2021-01-19 15:18:00               NaN   
999  Property Damage Crash 2021-01-19 15:18:00            County   

            road_name cross_street_type cross_street_name municipality  ...  \
0          NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
1          NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
2            RIDGE RD            County          BRINK RD          N/A  ...   
3            RIDGE RD            County          BRINK RD          N/A  ...   
4    W MONTGOMERY AVE      Municipality         NELSON ST    ROCKVILLE  ...   
..                ...               ...               ...          ...  ...   
995   WATKINS MILL RD            County    APPLE RIDGE RD          N/A  ...   
996         ELDRID DR            County       JOHNSON AVE          N/A  ...   
997         FENTON ST  Maryland (State)  PHILADELPHIA AVE          N/A  ...   
998               NaN               NaN               NaN          NaN  ...   
999         FENTON ST  Maryland (State)  PHILADELPHIA AVE          N/A  ...   

    speed_limit parked_vehicle vehicle_year vehicle_make      vehicle_model  \
0            50             No         2006       TOYOTA                 4S   
1            50             No         2014         TOYT                 4S   
2            40             No         2016         FORD              F-150   
3            40             No         2016       TOYOTA             SIENNA   
4            30             No         2015          BMW                M4S   
..          ...            ...          ...          ...                ...   
995          35             No         2010       TOYOTA              PRIUS   
996          35             No         2000         FORD  EXPLORER (POSTAL)   
997          35             No         2017      HYUNDAI            ELANTRA   
998          15             No         2001     MERCEDES              4DOOR   
999          35             No         2011         FORD             FUSION   

        latitude     longitude  \
0    39.09043867  -77.12867933   
1    39.09043867  -77.12867933   
2    39.21003667      -77.2399   
3    39.21003667      -77.2399   
4    39.08605617  -77.17062833   
..           ...           ...   
995  39.18419833  -77.21089333   
996     39.06622  -76.99723667   
997  38.98793019   -77.0235645   
998    39.000365  -77.00984667   
999  38.98793019   -77.0235645   

                                           geolocation related_non_motorist  \
0    {'l